### DISCUSSION TITLE
Strange reportPrivateImportUsage Error

### DISCUSSION THREAD

**Discussion by martinezlc99** [microsoft/pyright#9260] [Q&A] [answered]
Not sure if this behavior is expected, but getting Pylance `reportPrivateImportUsage` when "public "symbols are imported from an `__init__.py` which imports the symbols surrounded by parentheses. 

However, [removing the parenthesis around the `robot.api.parsing` symbols](https://github.com/robotframework/robotframework/blob/84711a7b6617e1150fe401366a26176c38d54af2/src/robot/api/__init__.py#L83) removes the error.

Concretely, updating this:

```python
from robot.parsing import (get_tokens, get_resource_tokens, get_init_tokens,
                           get_model, get_resource_model, get_init_model, Token)
```

to

```python
from robot.parsing import get_tokens, get_resource_tokens, get_init_tokens, get_model, get_resource_model, get_init_model, 
````

removes the error.  Any ideas why this is?

`Pylance` extension version 2024.10.1
`python` version 3.12.7
`macos` Sequoia 15.0.1

**Comment by erictraut** [comment]
You didn't provide the name of the package or the version that you've installed. I'm guessing that it's robotframework 7.1.1?

I'm not able to repro the problem as reported, so there must be additional steps required for a repro. Here's what I see when using the latest version of Pylance.

![image](https://github.com/user-attachments/assets/a929ab88-d2bf-430a-a051-45dd0cfde5f0)

**Reply by martinezlc99** [reply]
Apologies.  Yes, I am using robotframework 7.1.  

```python
from robot.api.parsing import get_model # and other symbols
```

results in the error.  Directly importing from `robot.parsing` as you do does not exhibit the issue as you observed.

**Reply by martinezlc99** [reply]
Just to be clear, updating the [`robot.api.__init__.py`](https://github.com/robotframework/robotframework/blob/84711a7b6617e1150fe401366a26176c38d54af2/src/robot/api/__init__.py#L83) to remove the parentheses around the imports removes the error.

I really appreciate your help here.

**Reply by erictraut** [reply, marked as answer]
This package is a "py.typed" library, which means standard-compliant type checkers like pyright follow [these rules](https://typing.readthedocs.io/en/latest/spec/distributing.html#library-interface-public-and-private-symbols) when determining whether a symbol is public or private.

In this case, the `get_model` symbol is not publicly exported from the `robot.api.parsing` submodule. If the library authors had intended for it to be publicly exported from this submodule, they would have needed to use a redundant import form (`from robot.parsing import get_model as get_model`) or include "get_model" in an `__all__` definition.

I'm not able to reproduce a change in behavior by removing parentheses around the imports in `robot.api.__init__.py`. I agree that the presence or absence of parens should not affect this behavior, so I'd consider this a bug. But I'd need a concrete repro to investigate it further. It's not clear to me why any change to that particular file would change the behavior in this case because it shouldn't be involved. 

The statement `from robot.api.parsing import get_model` attempts to import the symbol from the file `robot/api/parsing.py`. This file imports (but does not publicly re-export) the symbol from the file `robot/parsing/__init__.py`. This file imports (but does not publicly re-export) the symbol from `robot/parsing/parser/__init__.py`. This file imports (but does not publicly re-export) the symbol from `robot/parsing/parser/parser.py`, which actually defines a function by that name. The file `robot/api/__init__.py` shouldn't matter here because it's not involved in the import resolution for the statement `from robot.api.parsing import get_model`.

**Reply by martinezlc99** [reply]
Thank you Eric.  I am not sure whats going on, but you are correct that removing the parentheses from `robot.api.__init__.py` does not immediately resolve the issue, nor is it even involved in the import as you as mention.  But for some reason, when vscode is restarted the issue is resolved.  Re-adding the parentheses also recreates the error, but only after a vscode restart.

Not sure if a red herring, but I'll work on a repeatable example and get back to you if I can find a reproducible example.

Again, thank you for your time.
